alg/sha256.c

```diff
@@ -196,13 +194,17 @@ SHA256_Transform(uint32_t state[static restrict 8],
 {
 	int i;
 
+#ifdef HWACCEL
+	switch(usehw()) {
 #if defined(CPUSUPPORT_X86_SHANI) && defined(CPUSUPPORT_X86_SSSE3)
-	/* Use SHANI if we can. */
-	if (usehw()) {
+	case 1:
 		SHA256_Transform_shani(state, block);
 		return;
-	}
 #endif
+	default:
```

Comment:
Does anything change if we make this
```
0:
    break;
default:
    /* UNREACHABLE */
    assert(0);
```
in order to make it clear that we should be covering all of the same cases as `usehw` can emit?

Reply:
Change performance-wise, I mean.  Sometimes compilers can make smarter deductions about reachability given hints like this; sometimes they get confused and do stupid things.

If there's no performance difference I'd prefer to have the "default" case to go a failed assertion, just as a safeguard against future bugs.

Reply:
Yes, it makes a difference... in a very unpleasant manner.  It's all over the map!  :(

![compilers](https://user-images.githubusercontent.com/222157/105282959-52445800-5b64-11eb-9854-f69ec940ec81.png)

My reading:
- for `clang`, both `-O2` and `-O3 -march=native`, the best option is the current PR: `warn0(); break;`.  Note that we only expect this to ever occur during development, so maybe a warning message instead of an `abort()` is ok?
- for `gcc -O3 -march=native`, the best option is the same as `clang`, although only slightly.
- for `gcc -O2`, the best speed comes from `warn0(); abort()`.  That gives a +5% speed improvement over the baseline (and over `warn0(); break;`).
- in general, I'm slightly dismayed that there's *this* much variation in the speed.  I'm going to test using a callback function, instead of selecting an option this way.